pkg/server/commands/reverseexpand/reverse_expand_weighted.go

```diff
@@ -22,31 +21,44 @@ import (
 	"github.com/openfga/openfga/pkg/tuple"
 )
 
-var ErrEmptyStack = errors.New("unexpected empty stack")
+var ErrEmptyStack = errors.New("pop called on empty stack")
 
+// TypeRelEntry represents a step in the path taken to reach a leaf node.
 // As reverseExpand traverses from a requested type#rel to its leaf nodes, it pushes typeRelEntry structs to a stack.
-// Each `typeRelEntry` represents a step in the path taken to reach a leaf.
 // After reaching a leaf, this stack is consumed by the `queryForTuples` function to build the precise chain of
 // database queries needed to find the resulting objects.
-// To avoid races, every leaf node receives its own copy of the stack.
-type typeRelEntry struct {
+type TypeRelEntry struct {
 	typeRel string // e.g. "organization#admin"
 
 	// Only present for userset relations. Will be the userset relation string itself.
 	// For `rel admin: [team#member]`, usersetRelation is "member"
 	usersetRelation string
 }
 
-// When cloning requests to pass into separate goroutines, we need to create a copy of that
-// request's stack as well to avoid races.
-func cloneStack(original lls.Stack) lls.Stack {
-	newStack := lls.New()
+type Stack struct {
+	value TypeRelEntry
+	next  *Stack
+}
+
+func Push(s *Stack, entry TypeRelEntry) *Stack {
+	newStack := Stack{value: entry, next: nil}
+	newStack.next = s
+	return &newStack
+}
 
-	vals := original.Values()
-	for i := len(vals) - 1; i >= 0; i-- {
-		newStack.Push(vals[i])
+func Pop(s *Stack) (TypeRelEntry, *Stack, error) {
+	if IsEmpty(s) {
+		return TypeRelEntry{}, nil, ErrEmptyStack
 	}
-	return *newStack
+	return s.value, s.next, nil
+}
```

Comment:
Feels like these should be property of stack.  ie.,e 

func (s *Stack) Push(entry TypeRelEntry) *Stack {
...
}
and 
func (s *Stack) Pop() (TelRelEntry, error) {
}

Reply:
Similarly Peek and IsEmpty